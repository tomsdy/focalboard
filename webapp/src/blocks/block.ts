// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Utils} from '../utils'

const contentBlockTypes = ['text', 'image', 'divider', 'checkbox'] as const
const blockTypes = [...contentBlockTypes, 'board', 'view', 'card', 'comment'] as const
type ContentBlockTypes = typeof contentBlockTypes[number]
type BlockTypes = typeof blockTypes[number]

interface IBlock {
    readonly id: string
    readonly parentId: string
    readonly rootId: string
    readonly modifiedBy: string

    readonly schema: number
    readonly type: BlockTypes
    readonly title: string
    readonly fields: Readonly<Record<string, any>>

    readonly createAt: number
    readonly updateAt: number
    readonly deleteAt: number
}

interface IMutableBlock extends IBlock {
    id: string
    parentId: string
    rootId: string
    modifiedBy: string

    schema: number
    type: BlockTypes
    title: string
    fields: Record<string, any>

    createAt: number
    updateAt: number
    deleteAt: number
}

class MutableBlock implements IMutableBlock {
    id: string = Utils.createGuid()
    schema: number
    parentId: string
    rootId: string
    modifiedBy: string
    type: BlockTypes
    title: string
    fields: Record<string, any> = {}
    createAt: number = Date.now()
    updateAt = 0
    deleteAt = 0

    constructor(block: any = {}) {
        this.id = block.id || Utils.createGuid()
        this.schema = 1
        this.parentId = block.parentId || ''
        this.rootId = block.rootId || ''
        this.modifiedBy = block.modifiedBy || ''
        this.type = block.type || ''

        // Shallow copy here. Derived classes must make deep copies of their known properties in their constructors.
        this.fields = block.fields ? {...block.fields} : {}

        this.title = block.title || ''

        const now = Date.now()
        this.createAt = block.createAt || now
        this.updateAt = block.updateAt || now
        this.deleteAt = block.deleteAt || 0
    }
}

export type {ContentBlockTypes, BlockTypes}
export {blockTypes, contentBlockTypes, IBlock, IMutableBlock, MutableBlock}
