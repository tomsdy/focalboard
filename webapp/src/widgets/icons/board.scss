@use './standardIcon.scss';

.focalboard-body .BoardIcon {
    @extend .StandardIcon;
    stroke-width: 8px;
}
