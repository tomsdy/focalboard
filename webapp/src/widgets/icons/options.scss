.focalboard-body .OptionsIcon {
    color: rgba(var(--body-color), 0.5);
    font-size: 16px;
}
