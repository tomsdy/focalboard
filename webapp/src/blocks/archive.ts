// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IBlock} from './block'

interface IArchiveHeader {
    version: number
    date: number
}

interface IArchiveLine {
    type: string,
    data: any,
}

// This schema allows the expansion of additional line types in the future
interface IBlockArchiveLine extends IArchiveLine {
    type: 'block',
    data: IBlock
}

class ArchiveUtils {
    static buildBlockArchive(blocks: readonly IBlock[]): string {
        const header: IArchiveHeader = {
            version: 1,
            date: Date.now(),
        }

        const headerString = JSON.stringify(header)
        let content = headerString + '\n'
        for (const block of blocks) {
            const line: IBlockArchiveLine = {
                type: 'block',
                data: block,
            }
            const lineString = JSON.stringify(line)
            content += lineString
            content += '\n'
        }

        return content
    }

    static parseBlockArchive(contents: string): IBlock[] {
        const blocks: IBlock[] = []
        const allLineStrings = contents.split('\n')
        if (allLineStrings.length >= 2) {
            const headerString = allLineStrings[0]
            const header = JSON.parse(headerString) as IArchiveHeader
            if (header.date && header.version >= 1) {
                const lineStrings = allLineStrings.slice(1)
                let lineNum = 2
                for (const lineString of lineStrings) {
                    if (!lineString) {
                        // Ignore empty lines, e.g. last line
                        continue
                    }
                    const line = JSON.parse(lineString) as IArchiveLine
                    if (!line || !line.type || !line.data) {
                        throw new Error(`ERROR parsing line ${lineNum}`)
                    }
                    switch (line.type) {
                    case 'block': {
                        const blockLine = line as IBlockArchiveLine
                        const block = blockLine.data
                        blocks.push(block)
                        break
                    }
                    }

                    lineNum += 1
                }
            } else {
                throw new Error('ERROR parsing header')
            }
        }

        return blocks
    }
}

export {IArchiveHeader, IArchiveLine, IBlockArchiveLine, ArchiveUtils}
