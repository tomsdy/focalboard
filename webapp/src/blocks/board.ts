// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Utils} from '../utils'

import {IBlock, MutableBlock} from './block'

type PropertyType = 'text' | 'number' | 'select' | 'multiSelect' | 'date' | 'person' | 'file' | 'checkbox' | 'url' | 'email' | 'phone' | 'createdTime' | 'createdBy' | 'updatedTime' | 'updatedBy'

interface IPropertyOption {
    readonly id: string
    readonly value: string
    readonly color: string
}

interface IMutablePropertyOption {
    id: string
    value: string
    color: string
}

// A template for card properties attached to a board
interface IPropertyTemplate {
    readonly id: string
    readonly name: string
    readonly type: PropertyType
    readonly options: IPropertyOption[]
}

interface IMutablePropertyTemplate extends IPropertyTemplate {
    id: string
    name: string
    type: PropertyType
    options: IMutablePropertyOption[]
}

interface Board extends IBlock {
    readonly icon: string
    readonly description: string
    readonly showDescription: boolean
    readonly isTemplate: boolean
    readonly cardProperties: readonly IPropertyTemplate[]
    duplicate(): MutableBoard
}

class MutableBoard extends MutableBlock implements Board {
    get icon(): string {
        return this.fields.icon as string
    }
    set icon(value: string) {
        this.fields.icon = value
    }

    get description(): string {
        return this.fields.description as string
    }
    set description(value: string) {
        this.fields.description = value
    }

    get showDescription(): boolean {
        return Boolean(this.fields.showDescription)
    }
    set showDescription(value: boolean) {
        this.fields.showDescription = value
    }

    get isTemplate(): boolean {
        return Boolean(this.fields.isTemplate)
    }
    set isTemplate(value: boolean) {
        this.fields.isTemplate = value
    }

    get cardProperties(): IMutablePropertyTemplate[] {
        return this.fields.cardProperties as IPropertyTemplate[]
    }
    set cardProperties(value: IMutablePropertyTemplate[]) {
        this.fields.cardProperties = value
    }

    constructor(block: any = {}) {
        super(block)
        this.type = 'board'

        this.icon = block.fields?.icon || ''
        this.description = block.fields?.description || ''
        if (block.fields?.cardProperties) {
            // Deep clone of card properties and their options
            this.cardProperties = block.fields.cardProperties.map((o: IPropertyTemplate) => {
                return {
                    id: o.id,
                    name: o.name,
                    type: o.type,
                    options: o.options ? o.options.map((option) => ({...option})) : [],
                }
            })
        } else {
            this.cardProperties = []
        }
    }

    duplicate(): MutableBoard {
        const board = new MutableBoard(this)
        board.id = Utils.createGuid()
        return board
    }
}

export {Board, MutableBoard, PropertyType, IPropertyOption, IPropertyTemplate}
