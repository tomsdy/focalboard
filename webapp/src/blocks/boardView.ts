// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Utils} from '../utils'

import {IBlock, MutableBlock} from './block'
import {FilterGroup} from './filterGroup'

type IViewType = 'board' | 'table' | 'gallery' // | 'calendar' | 'list'
type ISortOption = { propertyId: '__title' | string, reversed: boolean }

interface BoardView extends IBlock {
    readonly viewType: IViewType
    readonly groupById?: string
    readonly sortOptions: readonly ISortOption[]
    readonly visiblePropertyIds: readonly string[]
    readonly visibleOptionIds: readonly string[]
    readonly hiddenOptionIds: readonly string[]
    readonly filter: FilterGroup
    readonly cardOrder: readonly string[]
    readonly columnWidths: Readonly<Record<string, number>>

    duplicate(): MutableBoardView
}

class MutableBoardView extends MutableBlock implements BoardView {
    get viewType(): IViewType {
        return this.fields.viewType
    }
    set viewType(value: IViewType) {
        this.fields.viewType = value
    }

    get groupById(): string | undefined {
        return this.fields.groupById
    }
    set groupById(value: string | undefined) {
        this.fields.groupById = value
    }

    get sortOptions(): ISortOption[] {
        return this.fields.sortOptions
    }
    set sortOptions(value: ISortOption[]) {
        this.fields.sortOptions = value
    }

    get visiblePropertyIds(): string[] {
        return this.fields.visiblePropertyIds
    }
    set visiblePropertyIds(value: string[]) {
        this.fields.visiblePropertyIds = value
    }

    get visibleOptionIds(): string[] {
        return this.fields.visibleOptionIds
    }
    set visibleOptionIds(value: string[]) {
        this.fields.visibleOptionIds = value
    }

    get hiddenOptionIds(): string[] {
        return this.fields.hiddenOptionIds
    }
    set hiddenOptionIds(value: string[]) {
        this.fields.hiddenOptionIds = value
    }

    get filter(): FilterGroup {
        return this.fields.filter
    }
    set filter(value: FilterGroup) {
        this.fields.filter = value
    }

    get cardOrder(): string[] {
        return this.fields.cardOrder
    }
    set cardOrder(value: string[]) {
        this.fields.cardOrder = value
    }

    get columnWidths(): Record<string, number> {
        return this.fields.columnWidths as Record<string, number>
    }
    set columnWidths(value: Record<string, number>) {
        this.fields.columnWidths = value
    }

    constructor(block: any = {}) {
        super(block)

        this.type = 'view'

        this.sortOptions = block.fields?.sortOptions?.map((o: ISortOption) => ({...o})) || []		// Deep clone
        this.visiblePropertyIds = block.fields?.visiblePropertyIds?.slice() || []
        this.visibleOptionIds = block.fields?.visibleOptionIds?.slice() || []
        this.hiddenOptionIds = block.fields?.hiddenOptionIds?.slice() || []
        this.filter = new FilterGroup(block.fields?.filter)
        this.cardOrder = block.fields?.cardOrder?.slice() || []
        this.columnWidths = {...(block.fields?.columnWidths || {})}

        if (!this.viewType) {
            this.viewType = 'board'
        }
    }

    duplicate(): MutableBoardView {
        const view = new MutableBoardView(this)
        view.id = Utils.createGuid()
        return view
    }
}

export {BoardView, MutableBoardView, IViewType, ISortOption}
