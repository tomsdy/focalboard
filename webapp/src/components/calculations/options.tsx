// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react'

import Select, {components} from 'react-select'

import {CSSObject} from '@emotion/serialize'

import {getSelectBaseStyle} from '../../theme'
import ChevronUp from '../../widgets/icons/chevronUp'
import {IPropertyTemplate} from '../../blocks/board'

type Option = {
    label: string
    value: string
    displayName: string
}

const Options:Record<string, Option> = {
    none: {value: 'none', label: 'None', displayName: 'Calculate'},
    count: {value: 'count', label: 'Count', displayName: 'Count'},
    countValue: {value: 'countValue', label: 'Count Value', displayName: 'Values'},
    countUniqueValue: {value: 'countUniqueValue', label: 'Count Unique Values', displayName: 'Unique'},
    sum: {value: 'sum', label: 'Sum', displayName: 'Sum'},
    average: {value: 'average', label: 'Average', displayName: 'Average'},
    median: {value: 'median', label: 'Median', displayName: 'Median'},
    min: {value: 'min', label: 'Min', displayName: 'Min'},
    max: {value: 'max', label: 'Max', displayName: 'Max'},
    range: {value: 'range', label: 'Range', displayName: 'Range'},
}

const optionsByType: Map<string, Option[]> = new Map([
    ['common', [Options.none, Options.count, Options.countValue, Options.countUniqueValue]],
    ['number', [Options.sum, Options.average, Options.median, Options.min, Options.max, Options.range]],
])

const baseStyles = getSelectBaseStyle()

const styles = {
    ...baseStyles,
    dropdownIndicator: (provided: CSSObject): CSSObject => ({
        ...baseStyles.dropdownIndicator(provided),
        pointerEvents: 'none',
    }),
    control: (): CSSObject => ({
        border: 0,
        width: '100%',
        margin: '0',
        display: 'flex',
        flexDirection: 'row',
    }),
    menu: (provided: CSSObject): CSSObject => ({
        ...provided,
        minWidth: '100%',
        width: 'max-content',
        background: 'rgb(var(--center-channel-bg-rgb))',
        right: '0',
        marginBottom: '0',
    }),
    singleValue: (provided: CSSObject): CSSObject => ({
        ...baseStyles.singleValue(provided),
        opacity: '0.8',
        fontSize: '12px',
        right: '0',
        textTransform: 'uppercase',
    }),
    valueContainer: (provided: CSSObject): CSSObject => ({
        ...baseStyles.valueContainer(provided),
        display: 'none',
        pointerEvents: 'none',
    }),
}

const DropdownIndicator = (props: any) => {
    return (
        <components.DropdownIndicator {...props}>
            <ChevronUp/>
        </components.DropdownIndicator>
    )
}

type Props = {
    value: string,
    menuOpen?: boolean
    onClose?: () => void
    onChange: (value: string) => void
    property: IPropertyTemplate
}

const CalculationOptions = (props: Props): JSX.Element => {
    const options = [...optionsByType.get('common')!]
    if (optionsByType.get(props.property.type)) {
        options.push(...optionsByType.get(props.property.type)!)
    }

    return (
        <Select
            styles={styles}
            value={Options[props.value]}
            isMulti={false}
            isClearable={true}
            name={'calculation_options'}
            className={'CalculationOptions'}
            options={options}
            menuPlacement={'auto'}
            isSearchable={false}
            components={{DropdownIndicator}}
            defaultMenuIsOpen={props.menuOpen}
            autoFocus={true}
            formatOptionLabel={(option: Option, meta) => {
                return meta.context === 'menu' ? option.label : option.displayName
            }}
            onMenuClose={() => {
                if (props.onClose) {
                    props.onClose()
                }
            }}
            onChange={(item) => {
                if (item?.value) {
                    props.onChange(item.value)
                }
            }}
        />
    )
}

export {
    CalculationOptions,
    Options,
    Option,
}
