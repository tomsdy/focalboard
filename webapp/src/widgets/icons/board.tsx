// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'

import './board.scss'

export default function BoardIcon(): JSX.Element {
    return (
        <svg
            xmlns='http://www.w3.org/2000/svg'
            className='BoardIcon Icon'
            viewBox='0 0 100 100'
        >
            <rect
                x='10'
                y='10'
                width='80'
                height='80'
                rx='5'
                ry='5'
            />
            <polyline
                points='28,25 28,55'
                style={{strokeWidth: '15px'}}
            />
            <polyline
                points='50,25 50,70'
                style={{strokeWidth: '15px'}}
            />
            <polyline
                points='72,25 72,45'
                style={{strokeWidth: '15px'}}
            />
        </svg>
    )
}
