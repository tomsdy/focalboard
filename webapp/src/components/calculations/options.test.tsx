// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {render} from '@testing-library/react'
import React from 'react'

import {IPropertyTemplate} from '../../blocks/board'

import {CalculationOptions} from './options'

describe('components/calculations/Options', () => {
    test('should match snapshot', () => {
        const property = {
            type: 'number',
        } as IPropertyTemplate

        const component = (
            <CalculationOptions
                value={'none'}
                onChange={() => {}}
                property={property}
            />
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot menu open', () => {
        const property = {
            type: 'number',
        } as IPropertyTemplate

        const component = (
            <CalculationOptions
                value={'none'}
                menuOpen={true}
                onChange={() => {}}
                property={property}
            />
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()
    })
})
