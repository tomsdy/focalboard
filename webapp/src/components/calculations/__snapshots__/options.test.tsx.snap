// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/calculations/Options should match snapshot 1`] = `
<div>
  <div
    class="CalculationOptions css-2b097c-container"
  >
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    >
      <span
        id="aria-selection"
      />
      <span
        id="aria-context"
      >
          Select is focused , press Down to open the menu, 
      </span>
    </span>
    <div
      class=" css-1s59geg-Control"
    >
      <div
        class=" css-1mxrbau-ValueContainer"
      >
        <div
          class=" css-1brck82-singleValue"
        >
          Calculate
        </div>
        <input
          aria-autocomplete="list"
          class="css-wmatm6-dummyInput-DummyInput"
          id="react-select-2-input"
          readonly=""
          tabindex="0"
          value=""
        />
      </div>
      <div
        class=" css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          class=" css-13eygzs-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
        <span
          class=" css-43ykx9-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class=" css-1f9iddu-indicatorContainer"
        >
          <i
            class="CompassIcon icon-chevron-up ChevronUpIcon"
          />
        </div>
      </div>
    </div>
    <input
      name="calculation_options"
      type="hidden"
      value="none"
    />
  </div>
</div>
`;

exports[`components/calculations/Options should match snapshot menu open 1`] = `
<div>
  <div
    class="CalculationOptions css-2b097c-container"
  >
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    >
      <span
        id="aria-selection"
      />
      <span
        id="aria-context"
      >
         10 results available. Use Up and Down to choose options, press Enter to select the currently focused option, press Escape to exit the menu, press Tab to select the option and exit the menu.
      </span>
    </span>
    <div
      class=" css-1s59geg-Control"
    >
      <div
        class=" css-1mxrbau-ValueContainer"
      >
        <div
          class=" css-1brck82-singleValue"
        >
          Calculate
        </div>
        <input
          aria-autocomplete="list"
          class="css-wmatm6-dummyInput-DummyInput"
          id="react-select-3-input"
          readonly=""
          tabindex="0"
          value=""
        />
      </div>
      <div
        class=" css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          class=" css-13eygzs-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
        <span
          class=" css-43ykx9-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class=" css-1f9iddu-indicatorContainer"
        >
          <i
            class="CompassIcon icon-chevron-up ChevronUpIcon"
          />
        </div>
      </div>
    </div>
    <div
      class=" css-1kslk4z-menu"
    >
      <div
        class=" css-g29tl0-MenuList"
      >
        <div
          class=" css-1cnkr6z-option"
          id="react-select-3-option-0"
          tabindex="-1"
        >
          None
        </div>
        <div
          class=" css-14xsrqy-option"
          id="react-select-3-option-1"
          tabindex="-1"
        >
          Count
        </div>
        <div
          class=" css-14xsrqy-option"
          id="react-select-3-option-2"
          tabindex="-1"
        >
          Count Value
        </div>
        <div
          class=" css-14xsrqy-option"
          id="react-select-3-option-3"
          tabindex="-1"
        >
          Count Unique Values
        </div>
        <div
          class=" css-14xsrqy-option"
          id="react-select-3-option-4"
          tabindex="-1"
        >
          Sum
        </div>
        <div
          class=" css-14xsrqy-option"
          id="react-select-3-option-5"
          tabindex="-1"
        >
          Average
        </div>
        <div
          class=" css-14xsrqy-option"
          id="react-select-3-option-6"
          tabindex="-1"
        >
          Median
        </div>
        <div
          class=" css-14xsrqy-option"
          id="react-select-3-option-7"
          tabindex="-1"
        >
          Min
        </div>
        <div
          class=" css-14xsrqy-option"
          id="react-select-3-option-8"
          tabindex="-1"
        >
          Max
        </div>
        <div
          class=" css-14xsrqy-option"
          id="react-select-3-option-9"
          tabindex="-1"
        >
          Range
        </div>
      </div>
    </div>
    <input
      name="calculation_options"
      type="hidden"
      value="none"
    />
  </div>
</div>
`;
